File: problem1.cpp
// 238. Product of Array Except Self
// Time Complexity : O(n)
// Space Complexity : O(1)
// Did this code successfully run on Leetcode : yes
// Three line explanation of solution in plain english

// firstly iterating through the input array backwards, calculating the product of elements from the 
// current position to the end and storing them in the result vector. and then iterating through the input array again,
// and updating the result vector by multiplying the product of elements before the current position with the 
//product of elements after the current position.


// Your code here along with comments explaining your approach
vector<int> productExceptSelf(vector<int>& nums) {
    int n = nums.size(), rp = 1; 
    vector<int> ans(n);
    ans[n-1] = nums[n-1];
    for(int i = n-2; i >= 0; i--)
        ans[i] = nums[i] * ans[i+1];
    for(int i = 0; i < n - 1; i++)
    {
        ans[i] = rp * ans[i+1];
        rp = rp*nums[i];
    }
    ans[n-1] = rp;
    return ans;
}

File: Problem_1.js
// Time Complexity : O(n)
// Space Complexity: O(1) as result array is not considered as taking space
// Did this code successfully run on Leetcode : Yes
// Three line explanation of solution in plain english
// Store product of left numbers at each index
// Multiply product of right numbers at each index with existing(left product)

// Your code here along with comments explaining your approach
// https://leetcode.com/problems/product-of-array-except-self/

/**
 * @param {number[]} nums
 * @return {number[]}
 */
var productExceptSelf = function (nums) {
    if (nums === null || nums.length === 0)
        return [];
    let rProd = 1;
    let ans = new Array(nums.length);
    // Store product of left numbers at each index
    ans[0] = 1;
    for (let i = 0; i < ans.length; i++) {
        ans[i] = rProd;
        rProd = rProd * nums[i];
    }
    // Multiply product of right numbers at each index with existing(left product)
    rProd = 1;
    for (let i = ans.length - 1; i >= 0; i--) {
        ans[i] *= rProd;
        rProd *= nums[i];
    }
    return ans;
};
